libs/x-content/src/main/java/org/elasticsearch/common/xcontent/NamedXContentRegistry.java

```diff
@@ -123,19 +123,18 @@ public NamedXContentRegistry(List<Entry> entries) {
         if (parsers == null) {
             if (registry.isEmpty()) {
                 // The "empty" registry will never work so we throw a better exception as a hint.
-                throw new NamedObjectNotFoundException("named objects are not supported for this parser");
+                throw new XContentParseException("named objects are not supported for this parser");
             }
-            throw new NamedObjectNotFoundException("unknown named object category [" + categoryClass.getName() + "]");
+            throw new XContentParseException("unknown named object category [" + categoryClass.getName() + "]");
         }
         Entry entry = parsers.get(name);
         if (entry == null) {
-            throw new NamedObjectNotFoundException(parser.getTokenLocation(), "unable to parse " + categoryClass.getSimpleName() +
-                " with name [" + name + "]: parser not found");
+            throw new NamedObjectNotFoundException(parser.getTokenLocation(), "unknown field [" + name + "]", parsers.keySet());
         }
         if (false == entry.name.match(name, parser.getDeprecationHandler())) {
             /* Note that this shouldn't happen because we already looked up the entry using the names but we need to call `match` anyway
              * because it is responsible for logging deprecation warnings. */
-            throw new NamedObjectNotFoundException(parser.getTokenLocation(),
+            throw new XContentParseException(parser.getTokenLocation(),
```

Comment:
All of the exceptions that I changed to `XContentParseException`  are really more development errors than usage errors.

Reply:
And they don't have the critical `candidates` set.